src/core/qgsfeaturefiltermodel.cpp

```diff
@@ -0,0 +1,466 @@
+#if QT_VERSION < QT_VERSION_CHECK(5, 6, 0)
+        Entry extraEntry = mEntries.takeAt( mExtraIdentifierValueIndex );
+        mEntries.prepend( extraEntry );
+#else
+        mEntries.move( mExtraIdentifierValueIndex, 0 );
+#endif
+        endMoveRows();
+      }
+      firstRow = 1;
+    }
+
+    QgsDebugMsg( QStringLiteral( "Entries 2: %1 " ).arg( mEntries.size() ) );
+
+    // Remove all entries (except for extra entry if existent)
+    beginRemoveRows( QModelIndex(), firstRow, mEntries.size() - firstRow );
+    mEntries.remove( firstRow, mEntries.size() - firstRow );
+    endRemoveRows();
+
+    QgsDebugMsg( QStringLiteral( "Entries 3: %1 " ).arg( mEntries.size() ) );
+
+    if ( currentEntryInNewList == -1 )
+    {
+      QgsDebugMsg( QStringLiteral( "Current value is NOT in new list" ) );
+      beginInsertRows( QModelIndex(), 1, entries.size() + 1 );
+      mEntries += entries;
+      endInsertRows();
+      setExtraIdentifierValueIndex( 0 );
+    }
+    else
+    {
+      QgsDebugMsg( QStringLiteral( "Current value is in new list" ) );
+      if ( currentEntryInNewList != 0 )
+      {
+        beginInsertRows( QModelIndex(), 0, currentEntryInNewList - 1 );
+        mEntries = entries.mid( 0, currentEntryInNewList );
+        endInsertRows();
+      }
+      else
+      {
+        mEntries.replace( 0, entries.at( 0 ) );
+      }
+
+      emit dataChanged( index( currentEntryInNewList, 0, QModelIndex() ), index( currentEntryInNewList, 0, QModelIndex() ) );
+
+      beginInsertRows( QModelIndex(), currentEntryInNewList + 1, newEntriesSize - currentEntryInNewList - 1 );
+      mEntries += entries.mid( currentEntryInNewList + 1 );
+      endInsertRows();
+      setExtraIdentifierValueIndex( currentEntryInNewList );
+    }
+
+    emit filterJobCompleted();
+  }
+  emit endUpdate();
+}
+
+void QgsFeatureFilterModel::gathererThreadFinished()
+{
+  delete mGatherer;
+  mGatherer = nullptr;
+  emit isLoadingChanged();
+}
+
+void QgsFeatureFilterModel::scheduledReload()
+{
+  if ( !mSourceLayer )
+    return;
+
+  bool wasLoading = false;
+
+  if ( mGatherer )
+  {
+    // Send the gatherer thread to the graveyard:
+    //   forget about it, tell it to stop and delete when finished
+    disconnect( mGatherer, &QgsFieldExpressionValuesGatherer::collectedValues, this, &QgsFeatureFilterModel::updateCompleter );
+    disconnect( mGatherer, &QgsFieldExpressionValuesGatherer::finished, this, &QgsFeatureFilterModel::gathererThreadFinished );
+    connect( mGatherer, &QgsFieldExpressionValuesGatherer::finished, mGatherer, &QgsFieldExpressionValuesGatherer::deleteLater );
+    mGatherer->stop();
+    wasLoading = true;
+  }
+
+  QgsFeatureRequest request;
+
+  if ( mShouldReloadCurrentFeature )
+  {
+    request.setFilterExpression( QStringLiteral( "%1 = %2" ).arg( QgsExpression::quotedColumnRef( mIdentifierField ), QgsExpression::quotedValue( mExtraIdentifierValue ) ) );
```

Comment:
Maybe `QgsExpression::createFieldEqualityExpression(...)` could be used in this case to manage null values.

Reply:
`NULL` values should not be used in the identifier field (primary key), so I'm not sure this extra check will have any value. Actually in database terms, the `createFieldEqualityExpression` method is flawed. NULL is not equal to NULL.

Reply:
Good point...